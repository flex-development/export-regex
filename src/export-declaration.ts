/**
 * @file EXPORT_DECLARATION_REGEX
 * @module export-regex/declaration
 */

/**
 * Declaration `export` statement regex. Ignores matches in comments.
 *
 * **Note**: Requires unicode support ([flag `u`][1]).
 *
 * [1]: https://javascript.info/regexp-unicode
 *
 * @see https://regex101.com/r/8HpMrA
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/export#syntax
 *
 * @example
 *  import { EXPORT_DECLARATION_REGEX } from '@flex-development/export-regex'
 *  import { dedent } from 'ts-dedent'
 *
 *  const code: string = dedent`
 *    export const ESM_SYNTAX_REGEX: RegExp = /pattern/gm;
 *    export declare const RESOLVE_EXTENSIONS: string[]
 *    export enum Snack{}
 *    export const enum Snack {}
 *    export interface User {}
 *    export class House {}
 *    export declare abstract class House extends Building {}
 *    export abstract class House {}
 *    export async function run() {}
 *    export async function* run() {}
 *    export function foo() {}
 *    export function* foo() {}
 *    export const $foo = 42
 *    export const $foo= 42
 *    export namespace Foo {}
 *    export type Foo = 'foo'
 *    export let name1, name2; // also var
 *    export const name1 = 1, name2 = 2; // also var, let
 *    export function functionName() {}
 *    export class ClassName {}
 *    export function* generatorFunctionName() {}
 *    export const { name1, name2: bar } = o;
 *    export const [ name1, name2 ] = array;
 *  `
 *
 *  const print = (matches: IterableIterator<RegExpMatchArray>): void => {
 *    console.debug([...matches].map(match => omit(match, ['input'])))
 *  }
 *
 *  print(code.matchAll(EXPORT_DECLARATION_REGEX))
 *  // [
 *  //   {
 *  //     '0': 'export const ESM_SYNTAX_REGEX',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': 'ESM_SYNTAX_REGEX',
 *  //     index: 0,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: 'ESM_SYNTAX_REGEX'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export declare const RESOLVE_EXTENSIONS',
 *  //     '1': 'declare',
 *  //     '2': 'const',
 *  //     '3': 'RESOLVE_EXTENSIONS',
 *  //     index: 53,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'declare',
 *  //       declaration: 'const',
 *  //       exports: 'RESOLVE_EXTENSIONS'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export enum Snack',
 *  //     '1': undefined,
 *  //     '2': 'enum',
 *  //     '3': 'Snack',
 *  //     index: 103,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'enum',
 *  //       exports: 'Snack'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const enum Snack',
 *  //     '1': undefined,
 *  //     '2': 'const enum',
 *  //     '3': 'Snack',
 *  //     index: 123,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const enum',
 *  //       exports: 'Snack'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export interface User',
 *  //     '1': undefined,
 *  //     '2': 'interface',
 *  //     '3': 'User',
 *  //     index: 150,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'interface',
 *  //       exports: 'User'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export class House',
 *  //     '1': undefined,
 *  //     '2': 'class',
 *  //     '3': 'House',
 *  //     index: 175,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'class',
 *  //       exports: 'House'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export declare abstract class House',
 *  //     '1': 'declare abstract',
 *  //     '2': 'class',
 *  //     '3': 'House',
 *  //     index: 197,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'declare abstract',
 *  //       declaration: 'class',
 *  //       exports: 'House'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export abstract class House',
 *  //     '1': 'abstract',
 *  //     '2': 'class',
 *  //     '3': 'House',
 *  //     index: 253,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'abstract',
 *  //       declaration: 'class',
 *  //       exports: 'House'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export async function run',
 *  //     '1': 'async',
 *  //     '2': 'function',
 *  //     '3': 'run',
 *  //     index: 284,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'async',
 *  //       declaration: 'function',
 *  //       exports: 'run'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export async function* run',
 *  //     '1': 'async',
 *  //     '2': 'function*',
 *  //     '3': 'run',
 *  //     index: 315,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: 'async',
 *  //       declaration: 'function*',
 *  //       exports: 'run'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function foo',
 *  //     '1': undefined,
 *  //     '2': 'function',
 *  //     '3': 'foo',
 *  //     index: 347,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function',
 *  //       exports: 'foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function* foo',
 *  //     '1': undefined,
 *  //     '2': 'function*',
 *  //     '3': 'foo',
 *  //     index: 372,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function*',
 *  //       exports: 'foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const $foo',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '$foo',
 *  //     index: 398,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '$foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const $foo',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '$foo',
 *  //     index: 421,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '$foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export namespace Foo',
 *  //     '1': undefined,
 *  //     '2': 'namespace',
 *  //     '3': 'Foo',
 *  //     index: 443,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'namespace',
 *  //       exports: 'Foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export type Foo',
 *  //     '1': undefined,
 *  //     '2': 'type',
 *  //     '3': 'Foo',
 *  //     index: 467,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'type',
 *  //       exports: 'Foo'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export let name1, name2',
 *  //     '1': undefined,
 *  //     '2': 'let',
 *  //     '3': 'name1, name2',
 *  //     index: 491,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'let',
 *  //       exports: 'name1, name2'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const name1 = 1, name2 = 2',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': 'name1 = 1, name2 = 2',
 *  //     index: 536,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: 'name1 = 1, name2 = 2'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function functionName',
 *  //     '1': undefined,
 *  //     '2': 'function',
 *  //     '3': 'functionName',
 *  //     index: 596,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function',
 *  //       exports: 'functionName'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export class ClassName',
 *  //     '1': undefined,
 *  //     '2': 'class',
 *  //     '3': 'ClassName',
 *  //     index: 639,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'class',
 *  //       exports: 'ClassName'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export function* generatorFunctionName',
 *  //     '1': undefined,
 *  //     '2': 'function*',
 *  //     '3': 'generatorFunctionName',
 *  //     index: 674,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'function*',
 *  //       exports: 'generatorFunctionName'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const { name1, name2: bar } = o',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '{ name1, name2: bar }',
 *  //     index: 727,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '{ name1, name2: bar }'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export const [ name1, name2 ] = array',
 *  //     '1': undefined,
 *  //     '2': 'const',
 *  //     '3': '[ name1, name2 ]',
 *  //     index: 767,
 *  //     groups: [Object: null prototype] {
 *  //       modifiers: undefined,
 *  //       declaration: 'const',
 *  //       exports: '[ name1, name2 ]'
 *  //     }
 *  //   }
 *  // ]
 *
 * @const {RegExp} EXPORT_DECLARATION_REGEX
 */
const EXPORT_DECLARATION_REGEX: RegExp =
  /(?<=^|[\n;](?:[\t ]*(?:\w+ )?)?)export\s*(?<modifiers>(?:\s*declare|\s*abstract|\s*async)+)?\s*(?<declaration>class|const +enum|const|enum|function\*?|interface|let|namespace|type(?! *\{)|var)\s+(?<exports>(?:[$_\p{ID_Start}][$\u200C\u200D\p{ID_Continue}]*(?=[\s=:;/({])(?!.*?,))|(?:[\w\t\n\r .,:$'"=-]+)|(?:[{[][\w\t\n\r .,:$'"-]+[}\]]))(?:\s*=\s*[$_\p{ID_Start}][$\u200C\u200D\p{ID_Continue}]*)?/gu

export default EXPORT_DECLARATION_REGEX
