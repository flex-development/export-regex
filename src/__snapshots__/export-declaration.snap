// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:EXPORT_DECLARATION_REGEX > exports > class > should match export abstract class [exports] 1`] = `
{
  "0": "export abstract class Pet",
  "1": "abstract",
  "2": "class",
  "3": "Pet",
  "groups": {
    "declaration": "class",
    "exports": "Pet",
    "modifiers": "abstract",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > class > should match export class [exports] 1`] = `
{
  "0": "export class Pet",
  "1": undefined,
  "2": "class",
  "3": "Pet",
  "groups": {
    "declaration": "class",
    "exports": "Pet",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > class > should match export declare abstract class [exports] 1`] = `
{
  "0": "export declare abstract class Pet",
  "1": "declare abstract",
  "2": "class",
  "3": "Pet",
  "groups": {
    "declaration": "class",
    "exports": "Pet",
    "modifiers": "declare abstract",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > const > should match export const [exports] 1`] = `
[
  {
    "0": "export const IGNORE_PATTERNS",
    "1": undefined,
    "2": "const",
    "3": "IGNORE_PATTERNS",
    "groups": {
      "declaration": "const",
      "exports": "IGNORE_PATTERNS",
      "modifiers": undefined,
    },
  },
  {
    "0": "export const { name1, name2: bar } = o",
    "1": undefined,
    "2": "const",
    "3": "{ name1, name2: bar }",
    "groups": {
      "declaration": "const",
      "exports": "{ name1, name2: bar }",
      "modifiers": undefined,
    },
  },
  {
    "0": "export const [ name1, name2 ] = array",
    "1": undefined,
    "2": "const",
    "3": "[ name1, name2 ]",
    "groups": {
      "declaration": "const",
      "exports": "[ name1, name2 ]",
      "modifiers": undefined,
    },
  },
]
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > const > should match export declare const [exports] 1`] = `
{
  "0": "export declare const IGNORE_PATTERNS",
  "1": "declare",
  "2": "const",
  "3": "IGNORE_PATTERNS",
  "groups": {
    "declaration": "const",
    "exports": "IGNORE_PATTERNS",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > const enum > should match export const enum [exports] 1`] = `
{
  "0": "export const enum Color",
  "1": undefined,
  "2": "const enum",
  "3": "Color",
  "groups": {
    "declaration": "const enum",
    "exports": "Color",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > const enum > should match export declare const enum [exports] 1`] = `
{
  "0": "export declare const enum Color",
  "1": "declare",
  "2": "const enum",
  "3": "Color",
  "groups": {
    "declaration": "const enum",
    "exports": "Color",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > enum > should match export declare enum [exports] 1`] = `
{
  "0": "export declare enum TokenKind",
  "1": "declare",
  "2": "enum",
  "3": "TokenKind",
  "groups": {
    "declaration": "enum",
    "exports": "TokenKind",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > enum > should match export enum [exports] 1`] = `
{
  "0": "export enum TokenKind",
  "1": undefined,
  "2": "enum",
  "3": "TokenKind",
  "groups": {
    "declaration": "enum",
    "exports": "TokenKind",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function > should match export async function [exports] 1`] = `
{
  "0": "export async function foo",
  "1": "async",
  "2": "function",
  "3": "foo",
  "groups": {
    "declaration": "function",
    "exports": "foo",
    "modifiers": "async",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function > should match export declare async function [exports] 1`] = `
{
  "0": "export async declare function foo",
  "1": "async declare",
  "2": "function",
  "3": "foo",
  "groups": {
    "declaration": "function",
    "exports": "foo",
    "modifiers": "async declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function > should match export declare function [exports] 1`] = `
{
  "0": "export declare function foo",
  "1": "declare",
  "2": "function",
  "3": "foo",
  "groups": {
    "declaration": "function",
    "exports": "foo",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function > should match export function [exports] 1`] = `
{
  "0": "export function foo",
  "1": undefined,
  "2": "function",
  "3": "foo",
  "groups": {
    "declaration": "function",
    "exports": "foo",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function* > should match export async function* [exports] 1`] = `
{
  "0": "export async function* generator",
  "1": "async",
  "2": "function*",
  "3": "generator",
  "groups": {
    "declaration": "function*",
    "exports": "generator",
    "modifiers": "async",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function* > should match export declare async function* [exports] 1`] = `
{
  "0": "export async declare function* generator",
  "1": "async declare",
  "2": "function*",
  "3": "generator",
  "groups": {
    "declaration": "function*",
    "exports": "generator",
    "modifiers": "async declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function* > should match export declare function* [exports] 1`] = `
{
  "0": "export declare function* generator",
  "1": "declare",
  "2": "function*",
  "3": "generator",
  "groups": {
    "declaration": "function*",
    "exports": "generator",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > function* > should match export function* [exports] 1`] = `
{
  "0": "export function* generator",
  "1": undefined,
  "2": "function*",
  "3": "generator",
  "groups": {
    "declaration": "function*",
    "exports": "generator",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > interface > should match export declare interface [exports] 1`] = `
{
  "0": "export declare interface IUser",
  "1": "declare",
  "2": "interface",
  "3": "IUser",
  "groups": {
    "declaration": "interface",
    "exports": "IUser",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > interface > should match export interface [exports] 1`] = `
{
  "0": "export interface IUser",
  "1": undefined,
  "2": "interface",
  "3": "IUser",
  "groups": {
    "declaration": "interface",
    "exports": "IUser",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > let > should match export declare let [exports] 1`] = `
{
  "0": "export declare let name1, name2",
  "1": "declare",
  "2": "let",
  "3": "name1, name2",
  "groups": {
    "declaration": "let",
    "exports": "name1, name2",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > let > should match export let [exports] 1`] = `
{
  "0": "export let name1, name2",
  "1": undefined,
  "2": "let",
  "3": "name1, name2",
  "groups": {
    "declaration": "let",
    "exports": "name1, name2",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > namespace > should match export declare namespace [exports] 1`] = `
{
  "0": "export declare namespace Utils",
  "1": "declare",
  "2": "namespace",
  "3": "Utils",
  "groups": {
    "declaration": "namespace",
    "exports": "Utils",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > namespace > should match export namespace [exports] 1`] = `
{
  "0": "export namespace Utils",
  "1": undefined,
  "2": "namespace",
  "3": "Utils",
  "groups": {
    "declaration": "namespace",
    "exports": "Utils",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > should match nested export statement(s) in declaration file 1`] = `
[
  {
    "0": "export declare abstract class Building",
    "1": "declare abstract",
    "2": "class",
    "3": "Building",
    "groups": {
      "declaration": "class",
      "exports": "Building",
      "modifiers": "declare abstract",
    },
  },
  {
    "0": "export declare class House",
    "1": "declare",
    "2": "class",
    "3": "House",
    "groups": {
      "declaration": "class",
      "exports": "House",
      "modifiers": "declare",
    },
  },
]
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > type > should match export declare type [exports] 1`] = `
{
  "0": "export declare type Options",
  "1": "declare",
  "2": "type",
  "3": "Options",
  "groups": {
    "declaration": "type",
    "exports": "Options",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > type > should match export type [exports] 1`] = `
{
  "0": "export type Options",
  "1": undefined,
  "2": "type",
  "3": "Options",
  "groups": {
    "declaration": "type",
    "exports": "Options",
    "modifiers": undefined,
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > var > should match export declare var [exports] 1`] = `
{
  "0": "export declare var name1, name2",
  "1": "declare",
  "2": "var",
  "3": "name1, name2",
  "groups": {
    "declaration": "var",
    "exports": "name1, name2",
    "modifiers": "declare",
  },
}
`;

exports[`unit:EXPORT_DECLARATION_REGEX > exports > var > should match export var [exports] 1`] = `
{
  "0": "export var name1, name2",
  "1": undefined,
  "2": "var",
  "3": "name1, name2",
  "groups": {
    "declaration": "var",
    "exports": "name1, name2",
    "modifiers": undefined,
  },
}
`;
